### PR TITLE
feat:support devServer.setupMiddlewares

### PR DESCRIPTION
**Pull request by 9aoy**
# PR Details

support devServer.setupMiddlewares & expose server.sockWrite method for external calls

this config is similar with [webpack#devServer.setupMiddlewares](https://webpack.js.org/configuration/dev-server/#devserversetupmiddlewares), but only expose limited server methods.

## Motivation and Context

The need from PIA

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/builder-doc/zh/config/tools/devServer.md [hunk 1]
```diff
@@ -170,6 +170,57 @@ export default {
...
+```js
+Array<
+  (
+    middlewares: {
+      unshift: (...handlers: RequestHandler[]) => void;
+      push: (...handlers: RequestHandler[]) => void;
+    },
+    server: {
+      sockWrite: (
+        type: string,
```
Comment (chenjiahan) [inline packages/builder/builder-doc/zh/config/tools/devServer.md:185]: 目前有哪些能用的 type 嘞，除了 'content-changed'

File: packages/server/server/src/server/dev-server.ts [hunk 2]
```diff
@@ -239,18 +274,18 @@ export class ModernDevServer extends ModernServer {
   }
 
   private async setupBeforeDevMiddleware() {
-    const { runner, conf } = this;
+    const { runner, conf, dev } = this;
 
-    const setupMids = conf.tools.devServer?.before || [];
+    const setupMids = dev.before || [];
```
Comment (chenjiahan) [inline packages/server/server/src/server/dev-server.ts:279]: 这里的改动不会影响现有的 modern.js 项目么

Comment (9aoy) [reply packages/server/server/src/server/dev-server.ts:279]: 全局搜了下，就 app-tools 里用到这个包，它会把 tools.devServer 的配置在 dev 上挂一份

image

File: packages/builder/builder-doc/zh/config/tools/devServer.md (hunk 1 above)
Comment (9aoy) [reply packages/builder/builder-doc/zh/config/tools/devServer.md:185]: sockWrite 本身是可以随便传的，现在是只处理 hash / still-ok / content-changed /  warnings / errors / invalid 这几种。需要加类型限制么？

File: packages/server/server/src/server/dev-server.ts (hunk 2 above)
Comment (chenjiahan) [reply packages/server/server/src/server/dev-server.ts:279]: okok，那没问题

File: packages/builder/builder-doc/zh/config/tools/devServer.md (hunk 1 above)
Comment (chenjiahan) [reply packages/builder/builder-doc/zh/config/tools/devServer.md:186]: 类型限制不用加哈，这几个类型如果有使用场景的话，可以在文档里写一下，如果没什么使用场景，就保持现状好了

**Review by chenjiahan — DISMISSED** [review]

**Comment by chenjiahan** [comment]
改了 Dev Server 要补一下 changeset 描述哈

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**